From 6a35765f5c6311982234f532bd388421905a53a0 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Fri, 14 Aug 2020 00:34:32 +0300
Subject: [PATCH 240/302] media: i2c: imx296: Fix power on

Increase the power count early when powering on to ensure controls get
applied.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
---
 drivers/media/i2c/imx296.c | 36 ++++++++++++++++++++----------------
 1 file changed, 20 insertions(+), 16 deletions(-)

diff --git a/drivers/media/i2c/imx296.c b/drivers/media/i2c/imx296.c
index d353fe89cb24..bee594e73b05 100644
--- a/drivers/media/i2c/imx296.c
+++ b/drivers/media/i2c/imx296.c
@@ -539,26 +539,30 @@ static int imx296_s_power(struct v4l2_subdev *sd, int on)
 
 	mutex_lock(imx->ctrls.lock);
 
-	if (imx->power_count == !on) {
-		if (on) {
-			ret = imx296_power_on(imx);
-			if (ret < 0)
-				goto done;
-			ret = imx296_setup(imx);
-			if (ret < 0) {
-				imx296_power_off(imx);
-				goto done;
-			}
-		} else {
-			imx296_power_off(imx);
-		}
-	}
-
-	/* Update the power count. */
+	/*
+	 * Update the power count. This must be done first, to ensure controls
+	 * will be applied by __v4l2_ctrl_handler_setup().
+	 */
 	imx->power_count += on ? 1 : -1;
 	WARN_ON(imx->power_count < 0);
 
+	if (imx->power_count == 1 && on) {
+		ret = imx296_power_on(imx);
+		if (ret < 0)
+			goto done;
+		ret = imx296_setup(imx);
+		if (ret < 0) {
+			imx296_power_off(imx);
+			goto done;
+		}
+	} else if (imx->power_count == 0 && !on) {
+		imx296_power_off(imx);
+	}
+
 done:
+	if (ret < 0)
+		imx->power_count -= on ? 1 : -1;
+
 	mutex_unlock(imx->ctrls.lock);
 	return ret;
 }
-- 
Regards,

Laurent Pinchart

